### PR TITLE
Add comprehensive input validation with detailed error responses

### PR DESCRIPTION
**Pull request by olaservo**
## Summary
- Added comprehensive input validation middleware for all Todo endpoints
- Implemented field-specific error messages with structured error responses  
- Added validation for title length, HTML tag prevention, and type checking
- Enhanced ID parameter validation for route security
- Added comprehensive test coverage for all validation scenarios

### New Files
- `src/middleware/validation.ts` - Core validation logic with custom error handling
- `src/middleware/validation.test.ts` - Comprehensive test suite (15+ test cases)

### Modified Files
- `src/routes/todos.ts` - Integrated validation middleware, improved error responses
- `README.md` - Added validation rules and error response format documentation
- `package.json` - Added supertest for API testing

### Validation Rules Implemented
- **Title**: Required, 1-100 chars, no HTML tags, string type validation
- **Description**: Optional, max 500 chars, no HTML tags, string type validation  
- **Completed**: Boolean type validation when provided
- **ID Parameters**: Positive integer validation for all routes

## Test Coverage
- Valid input acceptance
- Missing required fields
- Invalid data types
- Length limit enforcement
- HTML tag prevention (XSS protection)
- Multiple simultaneous errors
- Edge cases and boundary conditions

## API Response Changes
**Before**: Simple error strings
```json
{ "error": "Title is required" }
```

**After**: Structured error objects with field details
```json
{
  "error": "Validation failed", 
  "details": [
    { "field": "title", "message": "Title is required" },
    { "field": "description", "message": "Description must be 500 characters or less" }
  ]
}
```

## Security Improvements
- HTML tag prevention to mitigate XSS attacks
- Input sanitization with automatic trimming
- Type validation to prevent injection attacks
- ID format validation to prevent path traversal

## Testing Instructions

1. **Install dependencies**: `npm install`
2. **Run test suite**: `npm test` (should show 15+ passing tests)
3. **Test API manually**:
   ```bash
   npm run dev
   
   # Test valid creation
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"Valid Todo","description":"Valid description"}'
   
   # Test validation errors
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"","description":"'$(printf 'a%.0s' {1..501})'"}'
   ```

## Breaking Changes
**Error response format has changed** - API consumers will need to handle the new structured error format

## Future Enhancements
- [ ] Add rate limiting validation
- [ ] Implement custom validation rules configuration
- [ ] Add internationalization for error messages
- [ ] Consider adding request sanitization middleware

### CODE REVIEW

File: src/middleware/validation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    errors.push({ field: 'title', message: 'Title cannot contain HTML tags' });
+  }
+
+  // Validate description
+  if (description !== undefined && description !== null) {
+    if (typeof description !== 'string') {
+      errors.push({ field: 'description', message: 'Description must be a string', value: description });
+    } else if (description.length > 500) {
+      errors.push({ field: 'description', message: 'Description must be 500 characters or less' });
+    } else if (/<.*>/g.test(description)) {  // BUG: Inconsistent regex - allows <img> but title rejects it
```
Comment (dandeliongold) [inline src/middleware/validation.ts:42]: **Bug: Inconsistent HTML tag validation between title and description**

The regex patterns for HTML validation are inconsistent:
- Title: `/<[^>]*>/g` (matches any tag with content)
- Description: `/<.*>/g` (matches any content between angle brackets)

This means:
- `` would be rejected in title but allowed in description
- `<script>alert("xss")</script>` would be rejected in both

**Recommendation:** Use the same regex pattern for both fields, preferably the more restrictive one from title validation: `/<[^>]*>/g`

File: src/middleware/validation.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      details: errors
+    });
+  }
+
+  next();
+};
+
+export const validateTodoId = (req: Request, res: Response, next: NextFunction) => {
+  const { id } = req.params;
+
```
Comment (dandeliongold) [inline src/middleware/validation.ts:102]: **Bug: ID "0" validation issue**

The current regex `/^\d+$/` allows ID "0" which may not be valid depending on your ID generation strategy. Most systems use 1-based IDs.

**Recommendation:** Consider using `/^[1-9]\d*$/` to only allow positive integers starting from 1, or explicitly check for `parseInt(id) > 0` if you want to be more explicit.

File: src/routes/todos.ts [hunk 3]
```diff
@@ -9,54 +10,67 @@ router.get('/', (req: Request, res: Response) => {
...
+router.post('/', validateTodoCreate, (req: Request, res: Response) => {
   const { title, description }: CreateTodoRequest = req.body;
   
-  if (!title || title.trim() === '') {
-    return res.status(400).json({ error: 'Title is required' });
-  }
-  
-  const todo = todoStore.create({ title: title.trim(), description });
+  // BUG: No try-catch block - if todoStore.create throws, server will crash
+  const todo = todoStore.create({
```
Comment (dandeliongold) [inline src/routes/todos.ts:31]: **Bug: Missing error handling**

The `todoStore.create` call is not wrapped in a try-catch block. If the store throws an exception (e.g., due to internal errors, database connection issues, etc.), the server will crash.

**Recommendation:** Add proper error handling:
```typescript
try {
  const todo = todoStore.create({ 
    title: title.trim(), 
    description: description?.trim() || undefined 
  });
  res.status(201).json(todo);
} catch (error) {
  res.status(500).json({ 
    error: 'Internal server error',
    details: [{ field: 'general', message: 'Failed to create todo' }]
  });
}
```

File: src/middleware/validation.test.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        .post('/')
+        .send({ title: '<script>alert("xss")</script>' });
+
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'title',
+        message: 'Title cannot contain HTML tags'
+      });
+    });
+
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:61]: **Missing test coverage**

The comment indicates missing test coverage for description HTML validation. This is important since the regex patterns are inconsistent between title and description.

**Recommendation:** Add a test to verify description HTML validation:
```typescript
it('should reject description with HTML tags', async () => {
  const response = await request(app)
    .post('/')
    .send({ title: 'Valid Title', description: '<script>alert("xss")</script>' });

  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'description',
    message: 'Description cannot contain HTML tags'
  });
});
```

**Review by dandeliongold — CHANGES_REQUESTED** [review]
## Overall Assessment

This PR adds comprehensive input validation to the Todo API, which is a significant improvement for security and data integrity. The implementation follows good practices with structured error responses and comprehensive test coverage.

## ✅ Strengths

1. **Comprehensive validation rules** - Good coverage of common validation scenarios (required fields, length limits, type checking, HTML prevention)
2. **Structured error responses** - The new error format with field-specific details is much better than simple strings
3. **Security improvements** - HTML tag prevention helps mitigate XSS attacks
4. **Extensive test coverage** - 15+ test cases covering various scenarios
5. **Good documentation** - README updates clearly explain validation rules and error formats
6. **Proper middleware architecture** - Clean separation of concerns

## ⚠️ Issues Found

1. **Inconsistent HTML validation** - Different regex patterns between title and description fields
2. **ID validation edge case** - Allows ID "0" which may not be valid
3. **Missing error handling** - todoStore.create() not wrapped in try-catch
4. **Incomplete test coverage** - Missing tests for description HTML validation and ID edge cases

## 🔧 Recommendations

1. Fix the regex inconsistency in HTML validation
2. Add proper error handling in route handlers
3. Complete the missing test cases
4. Consider the ID "0" validation edge case

## Breaking Changes

The error response format change is properly documented. API consumers will need to update their error handling logic.

## Verdict

This is a solid improvement to the codebase. The issues found are minor and easily addressable. The security and validation improvements significantly enhance the API's robustness.
